<a href="https://colab.research.google.com/github/ElyorS/AI-application-system/blob/main/12204556_hw5_ai_application.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# Import TensorFlow and other libraries
import tensorflow as tf
import numpy as np
import matplotlib.pyplot as plt

# Load the MNIST dataset
mnist = tf.keras.datasets.mnist  # Corrected 'mist' to 'mnist'

# Split the dataset into training and test sets
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()  # Corrected 'mist' to 'mnist'

# Normalize the images
train_images = train_images / 255.0
test_images = test_images / 255.0

# Function to create and evaluate a model with different activation functions
def evaluate_activation(activation_function):
    model = tf.keras.Sequential([
        tf.keras.layers.Flatten(input_shape=(28, 28)),
        tf.keras.layers.Dense(128, activation=activation_function),
        tf.keras.layers.Dense(10, activation='softmax')
    ])

    model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

    model.fit(train_images, train_labels, epochs=5)
    test_loss, test_acc = model.evaluate(test_images, test_labels)

    print(f"Test accuracy with {activation_function}: {test_acc}")

# Evaluate models with different activation functions
activation_functions = ['sigmoid', 'tanh', 'relu']
for activation in activation_functions:
    evaluate_activation(activation)


11490434/11490434 [==============================] - 0s 0us/step
Epoch 1/5
1875/1875 [==============================] - 17s 9ms/step - loss: 0.3991 - accuracy: 0.8947
Epoch 2/5
1875/1875 [==============================] - 10s 5ms/step - loss: 0.1952 - accuracy: 0.9435
Epoch 3/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.1420 - accuracy: 0.9592
Epoch 4/5
1875/1875 [==============================] - 9s 5ms/step - loss: 0.1104 - accuracy: 0.9684
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.1002 - accuracy: 0.9683
Test accuracy with sigmoid: 0.9682999849319458
Epoch 1/5
1875/1875 [==============================] - 9s 5ms/step - loss: 0.2859 - accuracy: 0.9180
Epoch 2/5
1875/1875 [==============================] - 7s 4ms/step - loss: 0.1386 - accuracy: 0.9594
Epoch 3/5
1875/1875 [==============================] - 9s 5ms/step - loss: 0.0931 - accuracy: 0.9726
Epoch 4/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.067

***Activity 1***
----------------------

Test accuracy with sigmoid: 0.9682999849319458
**********************************************************
Test accuracy with tanh: 0.9742000102996826
**********************************************************
Test accuracy with relu: 0.9783999919891357
**********************************************************
All three activation functions (Sigmoid, Tanh, ReLU) resulted in models with relatively high test accuracies. The ReLU activation function performed the best in terms of accuracy, achieving nearly 97.84%, followed closely by Tanh with approximately 97.42%, and Sigmoid with approximately 96.83%. Additionally, ReLU and Tanh had similar training times per epoch, while Sigmoid took longer. These results indicate that ReLU and Tanh are well-suited for this classification task, while Sigmoid also performs reasonably well.


In [4]:
# Evaluate models with different learning rates
for Ir in [0.001, 0.01, 0.1]:
  optimizer = tf.keras.optimizers.Adam(learning_rate=Ir)
  model = tf.keras. Sequential ([
  tf.keras.layers. Flatten (input_shape=(28, 28)),
  tf.keras. layers. Dense (128, activation= 'relu'),
  tf.keras. layers. Dense (10, activation='softmax')
  ])
  model.compile(optimizer=optimizer, loss='sparse_categorical_crossentropy', metrics=['accuracy'])
  model.fit(train_images, train_labels, epochs=5, batch_size=32)
  test_loss, test_acc = model.evaluate(test_images, test_labels)
  print(f"Test accuracy with learning rate {Ir}: {test_acc}")

Epoch 1/5
1875/1875 [==============================] - 11s 6ms/step - loss: 0.2543 - accuracy: 0.9279
Epoch 2/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.1132 - accuracy: 0.9664
Epoch 3/5
1875/1875 [==============================] - 9s 5ms/step - loss: 0.0773 - accuracy: 0.9768
Epoch 4/5
1875/1875 [==============================] - 9s 5ms/step - loss: 0.0580 - accuracy: 0.9826
Epoch 5/5
313/313 [==============================] - 1s 3ms/step - loss: 0.0789 - accuracy: 0.9771
Test accuracy with learning rate 0.001: 0.9771000146865845
Epoch 1/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.2416 - accuracy: 0.9284
Epoch 2/5
1875/1875 [==============================] - 9s 5ms/step - loss: 0.1612 - accuracy: 0.9546
Epoch 3/5
1875/1875 [==============================] - 9s 5ms/step - loss: 0.1440 - accuracy: 0.9619
Epoch 4/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.1340 - accuracy: 0.9649
Epoch 5/5
313/313 [==============

In [5]:
# Evaluate models with different learning rates
for Ir in [0.001, 0.01, 0.1]:
  optimizer = tf.keras.optimizers.Adam(learning_rate=Ir)
  model = tf.keras. Sequential ([
  tf.keras.layers. Flatten (input_shape=(28, 28)),
  tf.keras. layers. Dense (128, activation= 'relu'),
  tf.keras. layers. Dense (10, activation='softmax')
  ])
  model.compile(optimizer=optimizer, loss='sparse_categorical_crossentropy', metrics=['accuracy'])
  model.fit(train_images, train_labels, epochs=5, batch_size=64)
  test_loss, test_acc = model.evaluate(test_images, test_labels)
  print(f"Test accuracy with learning rate {Ir}: {test_acc}")

Epoch 1/5
938/938 [==============================] - 5s 4ms/step - loss: 0.2927 - accuracy: 0.9186
Epoch 2/5
938/938 [==============================] - 4s 4ms/step - loss: 0.1286 - accuracy: 0.9629
Epoch 3/5
938/938 [==============================] - 7s 7ms/step - loss: 0.0887 - accuracy: 0.9743
Epoch 4/5
938/938 [==============================] - 4s 4ms/step - loss: 0.0671 - accuracy: 0.9797
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.0744 - accuracy: 0.9770
Test accuracy with learning rate 0.001: 0.9769999980926514
Epoch 1/5
938/938 [==============================] - 6s 5ms/step - loss: 0.2251 - accuracy: 0.9336
Epoch 2/5
938/938 [==============================] - 5s 5ms/step - loss: 0.1325 - accuracy: 0.9611
Epoch 3/5
938/938 [==============================] - 5s 5ms/step - loss: 0.1165 - accuracy: 0.9669
Epoch 4/5
938/938 [==============================] - 5s 6ms/step - loss: 0.1027 - accuracy: 0.9714
Epoch 5/5
313/313 [==============================]

In [6]:
# Evaluate models with different learning rates
for Ir in [0.001, 0.01, 0.1]:
  optimizer = tf.keras.optimizers.Adam(learning_rate=Ir)
  model = tf.keras. Sequential ([
  tf.keras.layers. Flatten (input_shape=(28, 28)),
  tf.keras. layers. Dense (128, activation= 'relu'),
  tf.keras. layers. Dense (10, activation='softmax')
  ])
  model.compile(optimizer=optimizer, loss='sparse_categorical_crossentropy', metrics=['accuracy'])
  model.fit(train_images, train_labels, epochs=5, batch_size=128)
  test_loss, test_acc = model.evaluate(test_images, test_labels)
  print(f"Test accuracy with learning rate {Ir}: {test_acc}")

Epoch 1/5
469/469 [==============================] - 3s 5ms/step - loss: 0.3611 - accuracy: 0.8999
Epoch 2/5
469/469 [==============================] - 3s 6ms/step - loss: 0.1647 - accuracy: 0.9530
Epoch 3/5
469/469 [==============================] - 4s 8ms/step - loss: 0.1162 - accuracy: 0.9665
Epoch 4/5
469/469 [==============================] - 2s 5ms/step - loss: 0.0911 - accuracy: 0.9740
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.0822 - accuracy: 0.9757
Test accuracy with learning rate 0.001: 0.9757000207901001
Epoch 1/5
469/469 [==============================] - 3s 5ms/step - loss: 0.2126 - accuracy: 0.9348
Epoch 2/5
469/469 [==============================] - 3s 6ms/step - loss: 0.1091 - accuracy: 0.9672
Epoch 3/5
469/469 [==============================] - 3s 7ms/step - loss: 0.0888 - accuracy: 0.9729
Epoch 4/5
469/469 [==============================] - 3s 6ms/step - loss: 0.0741 - accuracy: 0.9776
Epoch 5/5
313/313 [==============================]

***Activity 2***
---------------
**********************************************************
Batch size 32:
------------------------------

Test accuracy with learning rate 0.001: 0.9771000146865845

Test accuracy with learning rate 0.01: 0.9549000263214111

Test accuracy with learning rate 0.1: 0.60589998960495
**********************************************************
Batch size 64:
---------------------------

Test accuracy with learning rate 0.001: 0.9769999980926514

Test accuracy with learning rate 0.01: 0.9628000259399414

Test accuracy with learning rate 0.1: 0.7017999887466431
**********************************************************
Batch size 128:
-------------------

Test accuracy with learning rate 0.001: 0.9757000207901001

Test accuracy with learning rate 0.01: 0.9660000205039978

Test accuracy with learning rate 0.1: 0.8762000203132629
**********************************************************
 Smaller batch size with an appropriate learning rate (such as: 0.001) appears to be a good choice for achieving higher test accuracy on this dataset.


In [11]:
import tensorflow as tf

# Load the MNIST dataset
mnist = tf.keras.datasets.mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

# Normalize the images
train_images = train_images / 255.0
test_images = test_images / 255.0

# Baseline model
baseline_model = tf.keras.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28)),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dense(10, activation='softmax')
])

# Compile and train the baseline model
baseline_model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
baseline_model.fit(train_images, train_labels, epochs=5)

# Evaluate the baseline model
test_loss_baseline, test_acc_baseline = baseline_model.evaluate(test_images, test_labels)
print(f"Test accuracy for baseline model: {test_acc_baseline}")

# Model with Dropout
model_with_dropout = tf.keras.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28)),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dropout(0.5),  # Dropout layer with a dropout rate of 0.5
    tf.keras.layers.Dense(10, activation='softmax')
])

# Compile and train the model with dropout
model_with_dropout.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model_with_dropout.fit(train_images, train_labels, epochs=5)

# Evaluate the model with dropout
test_loss_dropout, test_acc_dropout = model_with_dropout.evaluate(test_images, test_labels)
print(f"Test accuracy for model with dropout: {test_acc_dropout}")

# Model with L2 Regularization
model_with_l2 = tf.keras.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28)),
    tf.keras.layers.Dense(128, activation='relu', kernel_regularizer=tf.keras.regularizers.l2(0.01)),
    tf.keras.layers.Dense(10, activation='softmax', kernel_regularizer=tf.keras.regularizers.l2(0.01))
])

# Compile and train the model with L2 regularization
model_with_l2.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model_with_l2.fit(train_images, train_labels, epochs=5)

# Evaluate the model with L2 regularization
test_loss_l2, test_acc_l2 = model_with_l2.evaluate(test_images, test_labels)
print(f"Test accuracy for model with L2 regularization: {test_acc_l2}")


Epoch 1/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.2537 - accuracy: 0.9291
Epoch 2/5
1875/1875 [==============================] - 11s 6ms/step - loss: 0.1121 - accuracy: 0.9676
Epoch 3/5
1875/1875 [==============================] - 10s 5ms/step - loss: 0.0774 - accuracy: 0.9766
Epoch 4/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0585 - accuracy: 0.9817
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.0744 - accuracy: 0.9771
Test accuracy for baseline model: 0.9771000146865845
Epoch 1/5
1875/1875 [==============================] - 10s 5ms/step - loss: 0.3936 - accuracy: 0.8848
Epoch 2/5
1875/1875 [==============================] - 9s 5ms/step - loss: 0.2175 - accuracy: 0.9351
Epoch 3/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.1804 - accuracy: 0.9469
Epoch 4/5
1875/1875 [==============================] - 10s 5ms/step - loss: 0.1636 - accuracy: 0.9506
Epoch 5/5
313/313 [=================

***Activity 3***
------------------
******************************************************
Test accuracy for baseline model: 0.9771000146865845
******************************************************
Test accuracy for model with dropout: 0.973800003528595
******************************************************
Test accuracy for model with L2 regularization: 0.9025999903678894
******************************************************************
Baseline model achieved the highest test accuracy, indicating that, for this specific task, it performed better than the models with dropout and L2 regularization. Dropout helped prevent overfitting to some extent but had a minor impact on accuracy. On the other hand, L2 regularization, while effective at reducing overfitting, resulted in a significant decrease in test accuracy. The choice of regularization technique should be based on the trade-off between preventing overfitting and maintaining model performance.

In [12]:
import tensorflow as tf
from sklearn.metrics import precision_score, recall_score, f1_score

# Assuming you have a trained model and test data
predictions = model.predict(test_images)

# Convert probability predictions to class labels
predicted_labels = np.argmax(predictions, axis=1)

# Calculate precision, recall, and F1-score using TensorFlow functions
precision = tf.keras.metrics.Precision()(test_labels, predicted_labels).numpy()
recall = tf.keras.metrics.Recall()(test_labels, predicted_labels).numpy()
f1 = 2 * (precision * recall) / (precision + recall)

print(f"Precision: {precision}")
print(f"Recall: {recall}")
print(f"F1-score: {f1}")


313/313 [==============================] - 2s 6ms/step
Precision: 0.9951300621032715
Recall: 0.9967849254608154
F1-score: 0.9959568252445301


***Activity 4***
---------------------
*******************************************************************
Precision - It's very high, around 99.51%. This means that when the model says something is positive, it's right nearly all the time, with very few mistakes.
*******************************************************************
Recall - Also very high, about 99.68%. This shows that the model is excellent at finding most of the positive things in the dataset, missing very few.
*******************************************************************
F1-Score - It's approximately 99.60%. This is a combination of precision and recall, showing that the model has a strong balance between being accurate when it predicts something as positive and not missing many positive things.

In [13]:
import tensorflow as tf
from tensorflow.keras import layers, models

# Load the MNIST dataset
mnist = tf.keras.datasets.mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

# Normalize the images
train_images = train_images.reshape((60000, 28, 28, 1))
test_images = test_images.reshape((10000, 28, 28, 1))
train_images, test_images = train_images / 255.0, test_images / 255.0

# Create the CNN model
model = models.Sequential([
   layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
   layers.MaxPooling2D((2, 2)),
   layers.Flatten(),
   layers.Dense(128, activation='relu'),
   layers.Dense(10, activation='softmax')
])

# Compile the CNN model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# Train the CNN model
model.fit(train_images, train_labels, epochs=5)

# Evaluate the CNN model on the test data
test_loss, test_acc = model.evaluate(test_images, test_labels)
print(f"Test accuracy for CNN: {test_acc}")

# Assuming you have a trained baseline feedforward neural network model
# Evaluate the baseline model on the test data
test_loss_baseline, test_acc_baseline = baseline_model.evaluate(test_images, test_labels)
print(f"Test accuracy for baseline feedforward NN: {test_acc_baseline}")


Epoch 1/5
1875/1875 [==============================] - 47s 24ms/step - loss: 0.1560 - accuracy: 0.9537
Epoch 2/5
1875/1875 [==============================] - 44s 23ms/step - loss: 0.0524 - accuracy: 0.9842
Epoch 3/5
1875/1875 [==============================] - 46s 25ms/step - loss: 0.0340 - accuracy: 0.9889
Epoch 4/5
1875/1875 [==============================] - 44s 24ms/step - loss: 0.0215 - accuracy: 0.9932
Epoch 5/5
313/313 [==============================] - 2s 7ms/step - loss: 0.0500 - accuracy: 0.9844
Test accuracy for CNN: 0.9843999743461609
313/313 [==============================] - 1s 2ms/step - loss: 0.0744 - accuracy: 0.9771
Test accuracy for baseline feedforward NN: 0.9771000146865845


***Activity 5 ***
---------------------
*************************************************************
Test accuracy for CNN: 0.9843999743461609
*************************************************************
Test accuracy for baseline feedforward neural netwrok model : 0.9771000146865845
*************************************************************
The CNN outperformed the baseline feedforward neural network in terms of accuracy, achieving a higher accuracy of 98.44% compared to 97.71%. However, the CNN required more time to train per epoch. Therefore, for this MNIST dataset, the CNN proved to be more effective in image classification.